subprojects/core/src/main/java/org/gradle/internal/vfs/EventDrivenFileWatcherRegistry.java

```diff
@@ -28,31 +27,30 @@
 import java.io.IOException;
 import java.nio.file.Paths;
 import java.util.Optional;
-import java.util.concurrent.ArrayBlockingQueue;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.function.Function;
 
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.CREATED;
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.INVALIDATED;
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.MODIFIED;
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.REMOVED;
 
-public abstract class AbstractEventDrivenFileWatcherRegistry implements FileWatcherRegistry {
-    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEventDrivenFileWatcherRegistry.class);
+public class EventDrivenFileWatcherRegistry implements FileWatcherRegistry {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventDrivenFileWatcherRegistry.class);
 
     private final FileWatcher watcher;
-    private final BlockingQueue<FileWatchEvent> fileEvents = new ArrayBlockingQueue<>(4096);
+    private final BlockingQueue<FileWatchEvent> fileEvents;
     private final Thread eventConsumerThread;
     private final AtomicReference<MutableFileWatchingStatistics> fileWatchingStatistics = new AtomicReference<>(new MutableFileWatchingStatistics());
     private final FileWatcherUpdater fileWatcherUpdater;
 
     private volatile boolean consumeEvents = true;
     private volatile boolean stopping = false;
 
-    public AbstractEventDrivenFileWatcherRegistry(FileWatcherCreator watcherCreator, ChangeHandler handler, Function<FileWatcher, FileWatcherUpdater> watcherUpdaterCreator) {
-        this.watcher = createWatcher(watcherCreator);
-        this.fileWatcherUpdater = watcherUpdaterCreator.apply(watcher);
+    public EventDrivenFileWatcherRegistry(FileWatcher watcher, ChangeHandler handler, FileWatcherUpdater fileWatcherUpdater, BlockingQueue<FileWatchEvent> fileEvents) {
+        this.watcher = watcher;
+        this.fileEvents = fileEvents;
+        this.fileWatcherUpdater = fileWatcherUpdater;
```

Comment:
This class now just keeps hold of this instance to return later. That doesn't seem good. I wonder if we could keep track of it in some better way. Or maybe it would be better if `FileWatcherRegistry` delegated to the updater. Maybe even wrap `FileWatcherRegistry` in an `MutableFileWatcherRegistry` to add the necessary methods.

Reply:
Currently, the watcher registry holds together various parts, which include the updater. If we want to move the updater out of here, then we need something else to hold it or to query the watcher from the registry. Both things don't seem very appealing to me.

Let's revisit this when we re-organized the project structure.